test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DerivedTypes/DerivedTypesTests.cs

```diff
@@ -0,0 +1,149 @@
+
+            Assert.Equal(expectedContent, await response.Content.ReadAsStringAsync());
+        }
+
+        [Fact]
+        public async Task RestrictEntityToDerivedTypeInstance_ForDerivedTypeNameBeforeKey()
+        {
+            // Name of the derived type preceeds key
+            string requestUri = this.BaseAddress + "/odata/Customers/Microsoft.Test.E2E.AspNet.OData.DerivedTypes.VipCustomer(2)";
+
+            var request = new HttpRequestMessage(HttpMethod.Get, requestUri);
+
+            var response = await this.Client.SendAsync(request);
+            Assert.True(response.IsSuccessStatusCode);
+
+            var expectedContent = string.Format("{{\"@odata.context\":\"{0}/odata/$metadata" +
+                "#Customers/Microsoft.Test.E2E.AspNet.OData.DerivedTypes.VipCustomer/$entity\"," +
+                "\"Id\":2,\"Name\":\"Customer 2\",\"LoyaltyCardNo\":\"9876543210\"}}", this.BaseAddress);
+
+            Assert.Equal(expectedContent, await response.Content.ReadAsStringAsync());
+        }
+
+        [Fact]
+        public async Task ReturnNotFound_ForKeyNotAssociatedWithDerivedType()
+        {
+            // Customer with Id 1 is not a VipCustomer
+            string requestUri = this.BaseAddress + "/odata/Customers(1)/Microsoft.Test.E2E.AspNet.OData.DerivedTypes.VipCustomer";
+
+            var request = new HttpRequestMessage(HttpMethod.Get, requestUri);
+
+            var response = await this.Client.SendAsync(request);
+            Assert.False(response.IsSuccessStatusCode);
+            Assert.Equal(System.Net.HttpStatusCode.NotFound, response.StatusCode);
+        }
+
+        [Fact]
+        public async Task RestrictEntitySetToDerivedTypeInstances_ThenExpandNavProperty()
+        {
+            string requestUri = this.BaseAddress + "/odata/Customers/Microsoft.Test.E2E.AspNet.OData.DerivedTypes.VipCustomer?$expand=Orders";
+
+            var request = new HttpRequestMessage(HttpMethod.Get, requestUri);
+
+            var response = await this.Client.SendAsync(request);
+            Assert.True(response.IsSuccessStatusCode);
+
+            var expectedContent = string.Format("{{\"@odata.context\":\"{0}/odata/$metadata" +
+                "#Customers/Microsoft.Test.E2E.AspNet.OData.DerivedTypes.VipCustomer(Orders())\"," +
+                "\"value\":[{{\"Id\":2,\"Name\":\"Customer 2\",\"LoyaltyCardNo\":\"9876543210\"," +
+                "\"Orders\":[{{\"Id\":2,\"Amount\":230}},{{\"Id\":3,\"Amount\":150}}]}}]}}", this.BaseAddress);
+
+            Assert.Equal(expectedContent, await response.Content.ReadAsStringAsync());
+        }
+
+        [Fact]
+        public async Task RestrictEntityToDerivedTypeInstance_ThenExpandNavProperty()
+        {
+            // Key preceeds name of the derived type
+            string requestUri = this.BaseAddress + "/odata/Customers(2)/Microsoft.Test.E2E.AspNet.OData.DerivedTypes.VipCustomer?$expand=Orders";
```

Comment:
> "/odata/Customers(2)/Microsoft.Test.E2E.AspNet.OData.DerivedTypes.VipCustomer?$expand=Orders" [](start = 50, length = 94)

can it work for use $expand after the "/.../cast/key"? Maybe a test to cover it? 
Expand this case to using [Theory]